_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+will come up in the review of the definition.
+Weird behavior, like `+` that does anything other than perform addition should raise a red flag,
+just like a function called `transfer()` that does not transfer anything.
+
+Reviewing the invocations of an operator should not pose significant problems either, as long as
+the reviewer is aware of the types involved, which is always necessary to understand the expression.
+For built-in types behavior cannot be altered at all.
+For UDVTs any operator must necessarily be user-defined so seeing it is not much different from
+seeing a function call.
+In both cases the name may be misleading and a look at the definition may be necessary.
+An operator can have only one definition, the definition must be explicitly bound in the file that
+defines the type and cannot be shadowed.
+
+### The Future of User-Defined Operators and UDVTs
+Now that we know the new feature and its limitations, the  question on everyone's mind is probably
+"Will it stay that way?".
+How will the feature evolve in the short and long term?
+
+#### Literal Suffixes
+The very next step for UDVTs is the ability to provide custom literals of user-defined value types.
+The ease of using literal values in expressions is a big part of what makes a type feel native.
+Solidity already provides fractional literals and our intention is to make them usable with custom
+fixed-point number implementations.
+To achieve that we are going to allow defining suffix functions that return UDVT by converting
+literals that are already available in the language:
+
+```solidity
+type UFixed16x2 is uint16;
+
+function uf(int16 mantissa, uint8 exponent) pure suffix returns (UFixed16x2) { ... }
+```
+```solidity
+UFixed16x2 circumference = 2 uf * 3.14 uf * radius;
+```
+
+While the application of a suffix will involve a function call, eventually, with the help of
+compile-time evaluation, literal suffixes are going to become as efficient as native literals.
+
+#### More Operators
+Many operators have been excluded from the initial implementation.
+Some of them will very likely be user-definable at some point.
+
+1. `<<`, `>>` and `**` were excluded due to their non-uniform arguments.
+    For built-in types, the right-hand side is always an unsigned type.
+1.  `!` was excluded because we have to decide whether it should return the same type as the argument
+    of just `bool`.
+    We are leaning towards the latter, but it needs more consideration.
+1. `++` and `--` are very convenient, and we want to have them eventually.
+    They just need to be handled in a special way, and there is more than way to do so.
+1. Composite assignment operators, such as `+=`, will likely be provided automatically
+    in the future as long as the corresponding arithmetic/bitwise operator is available.
+    Allowing a direct implementation is currently not impossible anyway, because there are no references
+    to value types in the language.
```

Comment:
I guess that was supposed to be "not possible" instead of "not impossible" :-)? Also generally a lot of conjecture again, we may also still decide to let them be freely implemented - but fine with keeping it here.

Reply:
> I guess that was supposed to be "not possible" instead of "not impossible" :-)?

Yes :)

Reply:
I know it's not certain, that's why it's not worded more strongly. It's not like we're promising this, it's just an overview of what *might* happen as far as we can tell today.